src/query/ts/m3db/encoded_step_iterator.go

```diff
@@ -0,0 +1,256 @@
+func (s *encodedStep) Values() []float64 { return s.values }
+
+func (it *encodedStepIter) Current() (block.Step, error) {
+	it.mu.RLock()
+	currentTime := it.currentTime
+	if !it.started || currentTime.After(it.bounds.End()) {
+		it.mu.RUnlock()
+		panic("out of bounds")
+	}
+
+	if it.err != nil {
+		it.mu.RUnlock()
+		return nil, it.err
+	}
+
+	step := &encodedStep{
+		time:   currentTime,
+		values: it.consolidator.consolidate(),
+	}
+
+	it.mu.RUnlock()
+	return step, nil
+}
+
+func (it *encodedStepIter) nextConsolidatedForStep(i int) bool {
+	peek := it.seriesPeek[i]
+	if peek.finished {
+		// No next value in this iterator
+		return false
+	}
+
+	nextTime := it.currentTime.Add(it.bounds.StepSize)
+	if peek.started {
+		point := peek.point
+		if point.Timestamp.After(it.currentTime) {
+			// This point exists further than the current step
+			// There are next values, but this point should be NaN
+			return true
+		}
+
+		// Currently at a potentially viable data point.
+		// Record previously peeked value, and all potentially valid
+		// values, then apply consolidation function to them to get the
+		// consolidated point.
+		it.consolidator.addPointForIterator(point, i)
+		// clear peeked point.
+		it.seriesPeek[i].started = false
+		// If at boundary, add the point as the current value.
+		if point.Timestamp.Equal(it.currentTime) {
+			return true
+		}
+	}
+
+	iter := it.seriesIters[i]
+	// Read through iterator until finding a data point outside of the
+	// range of this consolidated step; then consolidate those points into
+	// a value, set the next peek value, and return true.
+	for iter.Next() {
+		dp, _, _ := iter.Current()
+
+		// If this datapoint is before the current timestamp, add it as a
+		// consolidation candidate.
+		if !dp.Timestamp.After(it.currentTime) {
+			it.seriesPeek[i].started = false
+			it.consolidator.addPointForIterator(dp, i)
+		} else {
+			// This point exists further than the current step.
+			// Set peeked value to this point, then consolidate the retrieved
+			// series.
+			it.seriesPeek[i].point = dp
+			it.seriesPeek[i].started = true
+			return true
+		}
+
+		if err := iter.Err(); err != nil {
+			it.err = err
+		}
+	}
+
+	// Consolidate here if there is a point expected after this one.
+	if it.lastBlock && it.bounds.End().Equal(nextTime) ||
+		nextTime.After(it.bounds.End()) {
+		return false
+	}
+
+	return true
+}
+
+func (it *encodedStepIter) nextConsolidated() {
+	end := it.bounds.End()
+	// Check that current time is not before end since end is exclusive
+	if !it.currentTime.Before(end) {
+		it.exhausted = true
+		return
+	}
+
+	var anyNext bool
+	for i := range it.seriesIters {
+		if it.nextConsolidatedForStep(i) {
+			anyNext = true
+		}
+	}
+
+	if !anyNext {
+		it.exhausted = true
+	}
+}
+
+// Need to run an initial step; if there are any values
+// that appear at exactly the start, they must be added.
+func (it *encodedStepIter) initialStep() {
```

Comment:
not super clear if this is needed as a special case, will go through it again

Reply:
Currently it inits the seriesPeek here. This will become more of a special case on the next phase, so would prefer to keep it separate.